src/math/dct.c

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+
+#include <sof/audio/format.h>
+#include <sof/math/matrix.h>
+#include <sof/math/dct.h>
+#include <sof/math/sqrt.h>
+#include <sof/math/trig.h>
+#include <errno.h>
+#include <stdint.h>
+
+#define PI_Q29 Q_CONVERT_FLOAT(3.1415926536, 29)
+#define TWO_PI_Q24 Q_CONVERT_FLOAT(6.2831853072, 24)
+#define ONE_OVER_SQRT_TWO Q_CONVERT_FLOAT(0.7071067812, 31)
+#define HALF_Q1 Q_CONVERT_FLOAT(0.5, 1)
+#define TWO_Q29 Q_CONVERT_FLOAT(2, 29)
+
+/* Return a matrix for discrete cosine transform (DCT). The DCT type is hard
+ * coded to DCT-II with orthogonal matrix. See
+ * https://en.wikipedia.org/wiki/Discrete_cosine_transform#DCT-II
+ *
+ * The size is (num_in, num_out). Currently there are not other input parameters.
+ * TODO: Add option for no ortho norm.
+ */
+
+/**
+ * \brief Initialize a 16 bit DCT matrix. The actual DCT transform is a matrix
+ * multiply with the returned matrix.
+ * \param[in,out]  dct  In input provide DCT type and size, in output the DCT matrix
+ */
+int dct_initialize_16(struct dct_plan_16 *dct)
+{
+	int16_t dct_val;
+	int32_t arg;
+	int32_t cos;
+	int32_t c1;
+	int16_t c2;
+	int16_t nk;
+	int n;
+	int m;
+	int k;
+
+	if (dct->type != DCT_II || dct->ortho != true)
+		return -EINVAL;
+
+	if (dct->num_in > DCT_MATRIX_SIZE_MAX || dct->num_out > DCT_MATRIX_SIZE_MAX)
+		return -EINVAL;
+
+	dct->matrix = mat_matrix_alloc_16b(dct->num_in, dct->num_out, 15);
+	if (!dct->matrix)
+		return -ENOMEM;
+
+	c1 = PI_Q29 / dct->num_in;
+	arg = Q_SHIFT_RND(TWO_Q29 / dct->num_in, 29, 12);
+	c2 = sqrt_int16(arg); /* Q4.12 */
+	for (n = 0; n < dct->num_in; n++) {
+		for (k = 0; k < dct->num_out; k++) {
+			nk = (Q_SHIFT_LEFT(n, 0, 1) + HALF_Q1) * Q_SHIFT_LEFT(k, 0, 1); /*Q14.2 */
```

Comment:
misalignment in calculation:
1. n, k, are int type.
2. result nk is int16,

how to make sure the result is not overflow?

Reply:
There's a limit with cosine argument for DCT size. I need to add a size check. If will also ensure this can't overflow.

Reply:
. With DCT_MATRIX_SIZE_MAX=42 the nk as flat is max 1701.5. As Q14.2 it is 6806 that is much less than 32767.

Int16_t nk would work up to DCT_MATRIX_SIZE_MAX=91. If there's later need for 128 or 256 bins DCT this would need to change to int32_t.

Reply:
Thanks